modules/auth_email/utils/link_assertions.ts

```diff
@@ -0,0 +1,101 @@
+import { RuntimeError, ScriptContext } from "../module.gen.ts";
+import {
+	IDENTITY_INFO_LINK,
+	IDENTITY_INFO_PASSWORD,
+	IDENTITY_INFO_PASSWORDLESS,
+} from "./provider.ts";
+
+export async function ensureNotAssociated(
+	ctx: ScriptContext,
+	providerInfo:
+		| typeof IDENTITY_INFO_LINK
+		| typeof IDENTITY_INFO_PASSWORDLESS
+		| typeof IDENTITY_INFO_PASSWORD,
+	email: string,
+	shouldRejectOnExistence: (
+		linked: { userId: string; userToken: string },
+	) => boolean | Promise<boolean>,
+) {
+	// Ensure that the email is not already associated with another account
+	let existingIdentity: { userToken: string };
+	try {
+		existingIdentity = await ctx.modules.identities.signIn({
+			info: providerInfo,
+			uniqueData: {
+				identifier: email,
+			},
+		});
+	} catch {
+		// If the email is not associated in this way, the "sign in" will error and put us here.
+		ctx.log.info(
+			"Email is confirmed to not be associated with another account thru this provider",
+			["email", email],
+			["provider", JSON.stringify(providerInfo)],
+		);
+		return;
+	}
+	// Email matches an existing identity using this provider
+	const existingUser = await ctx.modules.users.authenticateToken(
+		existingIdentity,
+	);
+
+	if (
+		await shouldRejectOnExistence({
+			userId: existingUser.userId,
+			userToken: existingIdentity.userToken,
+		})
+	) {
+		ctx.log.error(
+			"Email is already associated with another account",
+			["email", email],
+			["existingUser", existingUser.userId],
+			["existingUserToken", existingIdentity.userToken],
+			["provider", JSON.stringify(providerInfo)],
+		);
+
+		// Revoke the user token just in case
+		const { tokens } = await ctx.modules.tokens.fetchByToken({
+			tokens: [existingIdentity.userToken],
+		});
+		await ctx.modules.tokens.revoke({
+			tokenIds: tokens.map((token) => token.id),
+		});
+
+		// Reject the request because the email is already associated with some
+		// account in some incompatible way
+		throw new RuntimeError("email_in_use");
+	} else {
+		ctx.log.info(
+			"Email is already associated with an account, but that is okay in this case",
+			["email", email],
+			["existingUser", existingUser.userId],
+			["existingUserToken", existingIdentity.userToken],
+			["provider", JSON.stringify(providerInfo)],
+		);
+
+		// Revoke the user token just in case
+		const { tokens } = await ctx.modules.tokens.fetchByToken({
+			tokens: [existingIdentity.userToken],
+		});
+		await ctx.modules.tokens.revoke({
+			tokenIds: tokens.map((token) => token.id),
+		});
+	}
+}
+
+export async function ensureNotAssociatedAll(
```

Comment:
this is very very expensive, it does at least 3 round trips to the database before doing anything else in the request. we should expose a batch api on identities module for this.

Reply:
Will do. I'm wondering if we should be able to select matching identities by `identityType`, as this may be something to check on in the future.